Add Registration for non-stateful scatter_nd_min and scatter_nd_max. Fixes issue #20402

pratjosh9 opened the pull request:
Modified tensorflow/tensorflow/core/kernels/scatter_nd_op.cc to add registration for scatter_min and scatter_max ops.

pratjosh9 committed:
Add registration for non_stateful scatter_nd_min and scatter_nd_max

pratjosh9 committed:
Fix comment.

pratjosh9 committed:
Merge pull request #1 from pratjosh9/non_stateful_scatter_min_max

Non stateful scatter min max

alextp reviewed (changes requested):
Great!

Now we need to do a few more things before this works. The operations you're mentioning are still not defined; you need to go to the op registration files and add the ops there too. ScatterNdMax/Min and ResourceScatterNdMax/Min should live in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/state_ops.cc#L206 and TensorScatterMax/Min should live in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/array_ops.cc#L2999

Then you need to make APIdef files for these ops mimicking the existing ones; see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/base_api/api_def_ResourceScatterAdd.pbtxt for ResourceScatterAdd and https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/base_api/api_def_TensorScatterAdd.pbtxt for TensorScatterAdd.

At this point these operations are a part of TensorFlow's GraphDef, and we need to decide how they're accessible from python.

For the ScatterNdMax/Min, these ops should be hidden by making an apidef file in the python_api subdirectory with visibility:HIDDEN in it (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/python_api/api_def_ResourceScatterAdd.pbtxt ) while for the TensorScatterMin/Max ops we should only have the non-deprecated endpoint which calls it tensor_scatter_nd_max / min (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/api_def/python_api/api_def_TensorScatterAdd.pbtxt ; you'll need a subset of that).

Once all of that is done, you'll need to regenerate our API golden files by running the test in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/tests/api_compatibility_test.py with --update_goldens=True.

On top of that it'd be great to have a unit test for each of these four operations you're adding so we don't accidentally break them.

Thanks a lot for doing this!

pratjosh9 commented:
Great. I will start working on it as soon as I can.

alextp commented:
Great. Let me know if you need any more specific pointers. Searching for the names of the existing ops (TensorScatterAdd and ResourceScatterNdAdd) in the TF codebase will mostly point you to the right places to change (except you'll also catch a few generated files, like ops.pbtxt, if you do that)

pratjosh9 committed:
Add Registration for ResourceScatterNdMax/Min and ScatterNdMax/Min

pratjosh9 committed:
Add Registration for TensorScatterMax/Min

pratjosh9 commented:
I've created the .pbtxt files for ResourceScatterNdMax/Min and TensorScatterMax/Min. I'm having a bit trouble as to what to enter in the summary section in the files.

Eg: ResourceScatterNdMin.pbtxt file.

```
op {
  graph_op_name: "ResourceScatterNdMin"
  in_arg {
    name: "ref"
    description: <<END
A resource handle. Must be from a VarHandleOp.
END
  }
  in_arg {
    name: "indices"
    description: <<END
A Tensor. Must be one of the following types: int32, int64.
A tensor of indices into ref.
END
  }
  in_arg {
    name: "updates"
    description: <<END
A Tensor. Must have the same type as ref. A tensor of
values whose element wise min is taken with ref.
END
  }
  attr {
    name: "use_locking"
    description: <<END
An optional bool. Defaults to True. If True, the assignment will
be protected by a lock; otherwise the behavior is undefined,
but may exhibit less contention.
END
  }
```

Eg. TensorScatterMin.pbtxt
```
op {
  graph_op_name: "TensorScatterMin"
  in_arg {
    name: "tensor"
    description: <<END
Tensor to update.
END
  }
  in_arg {
    name: "indices"
    description: <<END
Index tensor.
END
  }
  in_arg {
    name: "updates"
    description: <<END
Updates to scatter into output.
END
  }
  out_arg {
    name: "output"
    description: <<END
A new tensor copied from tensor whose values are element-wise minimum between 
tensor and updates according to the indices.
END
}
```

Also should both ScatterNdMax/Min and ResourceScatterMax/Min ops be hidden in the python API ?
And in which files should I write unit tests for these new ops?

alextp commented:
Re what to put in the summary section, for those hidden ops it doesn't matter too much.

Yes, scatter_nd_min and resource_scatter_nd_min should be hidden. Expose them by adding scatter_nd_max.min methods to ResourceVariable which call the ops (see how the other scatter methods there are implemented).

alextp commented:
Also the unit tests for the resource scatter can go in tensorflow/python/kernel_tests/scatter_nd_ops_test.py

pratjosh9 committed:
Remove Incorrect Registrations

pratjosh9 commented:
API golden files cannot be generated as tensorflow is giving the error 
File "api_compatibility_test.py", line 46, in <module>
    from tensorflow.tools.api.lib import api_objects_pb2
ModuleNotFoundError: No module named 'tensorflow.tools.api'

I've looked into similar issues and installed the latest release of tensorflow using pip but still this error is not resolving. Do I have to build tensorflow from source?

alextp commented:
You need to run the test with bazel run

On Wed, Apr 3, 2019 at 9:12 AM Pratik Joshi <notifications@github.com>
wrote:

> API golden files cannot be generated as tensorflow is giving the error
> File "api_compatibility_test.py", line 46, in
> from tensorflow.tools.api.lib import api_objects_pb2
> ModuleNotFoundError: No module named 'tensorflow.tools.api'
>
> I've looked into similar issues and installed the latest release of
> tensorflow using pip but still this error is not resolving. Do I have to
> build tensorflow from source?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/26923#issuecomment-479555508>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxfNs-OObxsmZFKY87pMhncBSrYoHks5vdNLkgaJpZM4b-o80>
> .
>


-- 
 - Alex

pratjosh9 commented:
Bazel build failed as Scatter_Nd_UpdateOp does not contain MAX and MIN definitions. I am defining them in tensorflow/tensorflow/core/kernels/dense_update_functor.cc and it's related files according to the definitions already and the ones in scatter_functor.cc.

alextp commented:
Ah makes sense. Thanks!

On Thu, Apr 4, 2019 at 10:43 PM Pratik Joshi <notifications@github.com>
wrote:

> Bazel build failed as Scatter_Nd_UpdateOp does not contain MAX and MIN
> definitions. I am defining them in
> tensorflow/tensorflow/core/kernels/dense_update_functor.cc and it's related
> files according to the definitions already and the ones in
> scatter_functor.cc.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/26923#issuecomment-480154553>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxUIhOaPSfdL2ANTIgvRM6GydABgmks5vduJvgaJpZM4b-o80>
> .
>


-- 
 - Alex

pratjosh9 commented:
I've implemented the changes but the build process is failing when I try to generate the api_golden files. Can you tell me how to resolve this error?
```
ERROR: /home/pj/Code/OpenSource/tensorflow/tensorflow/BUILD:713:1: Executing genrule //tensorflow:tf_python_api_gen_v1 failed (Exit 1): bash failed: error executing command 
  (cd /home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow && \
  exec env - \
    PATH=/home/pj/anaconda3/envs/tflow/bin:/home/pj/anaconda3/condabin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl \
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1 --root_init_template=tensorflow/api_template_v1.__init__.py --apidir=bazel-out/k8-opt/genfiles/tensorflow_api/v1/ --apiname=tensorflow --apiversion=1 --compat_apiversion=1 --compat_apiversion=2  --compat_init_template=tensorflow/compat_template_v1.__init__.py --compat_init_template=tensorflow/compat_template.__init__.py --package=tensorflow.python,tensorflow.lite.python.lite --output_package=tensorflow._api.v1 bazel-out/k8-opt/genfiles/tensorflow/_api/v1/v1.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/app/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/audio/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/autograph/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/autograph/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/bitwise/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/config/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/config/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/config/gpu/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/config/optimizer/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/config/threading/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/data/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/data/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/debugging/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/distribute/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/distribute/cluster_resolver/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/distribute/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/distributions/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/dtypes/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/errors/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/feature_column/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/gfile/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/io/gfile/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/graph_util/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/image/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/io/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/queue/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/initializers/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/layers/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/layers/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/linalg/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/lite/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/lite/constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/lite/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/lite/experimental/nn/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/logging/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/lookup/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/lookup/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/losses/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/manip/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/math/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/metrics/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/nest/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/nn/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/nn/rnn_cell/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/profiler/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/python_io/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/quantization/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/ragged/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/random/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/random/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/raw_ops/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/resource_loader/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/strings/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/builder/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/loader/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/main_op/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/signature_constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/signature_def_utils/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/tag_constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/saved_model/utils/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/sets/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/signal/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/sparse/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/spectral/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/summary/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/sysconfig/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/test/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/tpu/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/tpu/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/train/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/train/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/train/queue_runner/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/user_ops/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/version/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/audio/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/autograph/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/autograph/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/bitwise/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/compat/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/config/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/config/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/config/gpu/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/config/optimizer/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/config/threading/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/data/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/data/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/debugging/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/distribute/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/distribute/cluster_resolver/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/distribute/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/dtypes/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/errors/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/feature_column/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/io/gfile/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/graph_util/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/image/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/io/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/queue/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/linalg/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/lite/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/lite/constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/lite/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/lite/experimental/nn/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/lookup/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/lookup/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/math/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/nest/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/nn/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/quantization/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/ragged/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/random/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/random/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/raw_ops/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/saved_model/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/sets/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/signal/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/sparse/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/strings/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/summary/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/summary/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/sysconfig/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/test/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/tpu/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/tpu/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/train/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/train/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v2/version/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/app/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/audio/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/autograph/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/autograph/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/bitwise/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/compat/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/config/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/config/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/config/gpu/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/config/optimizer/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/config/threading/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/data/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/data/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/debugging/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/distribute/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/distribute/cluster_resolver/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/distribute/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/distributions/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/dtypes/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/errors/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/feature_column/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/gfile/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/io/gfile/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/graph_util/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/image/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/io/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/queue/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/initializers/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/layers/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/layers/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/linalg/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/lite/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/lite/constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/lite/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/lite/experimental/nn/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/logging/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/lookup/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/lookup/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/losses/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/manip/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/math/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/metrics/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/nest/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/nn/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/nn/rnn_cell/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/profiler/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/python_io/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/quantization/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/ragged/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/random/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/random/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/raw_ops/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/resource_loader/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/strings/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/builder/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/loader/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/main_op/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/signature_constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/signature_def_utils/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/tag_constants/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/saved_model/utils/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/sets/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/signal/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/sparse/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/spectral/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/summary/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/sysconfig/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/test/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/tpu/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/tpu/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/train/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/train/experimental/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/train/queue_runner/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/user_ops/__init__.py bazel-out/k8-opt/genfiles/tensorflow/_api/v1/compat/v1/version/__init__.py')
Execution platform: @bazel_tools//platforms:host_platform
Traceback (most recent call last):
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: /home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN10tensorflow7functor16ScatterNdFunctorIN5Eigen16ThreadPoolDeviceESt7complexIfExLNS_13scatter_nd_op8UpdateOpE3ELi7EEclERKS3_xNS2_5arrayIlLm7EEENS2_9TensorMapINS2_6TensorIS5_Li2ELi1ElEELi16ENS2_11MakePointerEEENSD_INSE_IKxLi2ELi1ElEELi16ESG_EENSD_INSE_IKS5_Li2ELi1ElEELi16ESG_EESH_

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/tools/api/generator/create_python_api.py", line 27, in <module>
    from tensorflow.python.tools.api.generator import doc_srcs
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 74, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: /home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/create_tensorflow.python_api_1_tf_python_api_gen_v1.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN10tensorflow7functor16ScatterNdFunctorIN5Eigen16ThreadPoolDeviceESt7complexIfExLNS_13scatter_nd_op8UpdateOpE3ELi7EEclERKS3_xNS2_5arrayIlLm7EEENS2_9TensorMapINS2_6TensorIS5_Li2ELi1ElEELi16ENS2_11MakePointerEEENSD_INSE_IKxLi2ELi1ElEELi16ESG_EENSD_INSE_IKS5_Li2ELi1ElEELi16ESG_EESH_
```

I've implemented the functors for the new functions as follows:
In **scatter_functor_gpu.cu.h**
```
template <typename T>
struct ScatterOpKernelBody<T, scatter_op::UpdateOp::MIN> {
  __device__ void operator()(T* dest, T src) const { CudaAtomicMin(dest, src); }
};

template <typename T>
struct ScatterOpKernelBody<T, scatter_op::UpdateOp::MAX> {
  __device__ void operator()(T* dest, T src) const { CudaAtomicMax(dest, src); }
};
```
In **scatter_nd_op_cpu_impl.h**
```
template <typename Input, typename Update, typename Output>
class UpdateExecutor<Input, Update, Output, scatter_nd_op::UpdateOp::MAX> {
 public:
  EIGEN_STRONG_INLINE static void Execute(Input  value, Update update,
                                          Output output) {
    output = value.cwiseMax(update);
  }
};

template <typename Input, typename Update, typename Output>
class UpdateExecutor<Input, Update, Output, scatter_nd_op::UpdateOp::MIN> {
 public:
  EIGEN_STRONG_INLINE static void Execute(Input value, Update update,
                                          Output output) {
    output = value.cwiseMin(update);
  }
};
```
In **scatter_nd_op_gpu.cc**
```
template <typename T>
struct LeftUpdate<T, scatter_nd_op::UpdateOp::MAX> {
  EIGEN_STRONG_INLINE EIGEN_DEVICE_FUNC void operator()(T* out, const T& val) {
    CudaAtomicMax(out, val);
  }
};

template <typename T>
struct LeftUpdate<T, scatter_nd_op::UpdateOp::SUB> {
  EIGEN_STRONG_INLINE EIGEN_DEVICE_FUNC void operator()(T* out, const T& val) {
    CudaAtomicMax(out, val);
  }
};
```

Earlier the build process was failing due to errors in these files, so I modified scatter_nd_op.cc and removed MAX/MIN definitions for complex types.

alextp commented:
@pratjosh9 you say you've made all these changes but they don't show up on the "files changed" tab of the PR.

Can you make sure you made the changes to the right branch?

pratjosh9 commented:
The changes are on my local system. Since I couldn't generate the api golden files due to the bazel error (mentioned above). I've not added the changed files to the PR. The changes I've made are mentioned above. 

Should I add changes (without api golden files) to the PR?

alextp commented:
Yes

On Mon, Apr 22, 2019 at 11:49 PM Pratik Joshi <notifications@github.com>
wrote:

> The changes are on my local system. Since I couldn't generate the api
> golden files due to the bazel error (mentioned above). I've not added the
> changed files to the PR. The changes I've made are mentioned above.
>
> Should I add changes (without api golden files) to the PR?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/26923#issuecomment-485664126>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABHRI6ZR3267I55V7BRJTPR2WN3ANCNFSM4G72R42A>
> .
>


-- 
 - Alex

pratjosh9 committed:
Add definition scatter_nd_min and scatter_nd_max.

pratjosh9 committed:
Add tests for scatter_nd_min and scatter_nd_max.

pratjosh9 committed:
Add api definitions for scatter_nd_min and scatter_nd_max.

pratjosh9 committed:
Merge pull request #2 from pratjosh9/non_stateful_scatter_min_max

Non stateful scatter min max

pratjosh9 commented:
Done

alextp commented on tensorflow/core/api_def/python_api/api_def_TensorScatterMax.pbtxt:
```diff
@@ -0,0 +1,11 @@
+op {
+  graph_op_name: "TensorScatterMax"
+  deprecation_message: "Use tensor_scatter_nd_max instead"
```
Why are you creating it with a deprecated endpoint already? Just add the new endpoint we want to keep.

alextp commented on tensorflow/core/ops/state_ops.cc:
```diff
@@ -260,6 +278,26 @@ REGISTER_OP("ScatterNdSub")
     .Attr("use_locking: bool = false")
     .SetShapeFn(shape_inference::ScatterNdUpdateShape);
 
+REGISTER_OP("ScatterNdMin")
```
I'd not bother registering these, as non-resource variables won't be really supported after tf2.0

alextp commented:
Do you get the bazel error when you do `bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens=True`?

If so, can you post the full stack trace here?

pratjosh9 commented:
> Do you get the bazel error when you do `bazel run tensorflow/tools/api/tests:api_compatibility_test -- --update_goldens=True`?
> 
> If so, can you post the full stack trace here?

I've removed the unnecessary changes and tried to generate the api_golden files. This is the error that I get. Can it be because of tensorflow being installed using Anaconda?

```
ERROR: /home/pj/Code/OpenSource/tensorflow/tensorflow/python/keras/api/BUILD:45:1: Executing genrule //tensorflow/python/keras/api:keras_python_api_gen_compat_v2 failed (Exit 1)
Traceback (most recent call last):
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: /home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN10tensorflow7functor16ScatterNdFunctorIN5Eigen16ThreadPoolDeviceESt7complexIfExLNS_13scatter_nd_op8UpdateOpE3ELi7EEclERKS3_xNS2_5arrayIlLm7EEENS2_9TensorMapINS2_6TensorIS5_Li2ELi1ElEELi16ENS2_11MakePointerEEENSD_INSE_IKxLi2ELi1ElEELi16ESG_EENSD_INSE_IKS5_Li2ELi1ElEELi16ESG_EESH_

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/tools/api/generator/create_python_api.py", line 27, in <module>
    from tensorflow.python.tools.api.generator import doc_srcs
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 74, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 58, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 28, in <module>
    _pywrap_tensorflow_internal = swig_import_helper()
  File "/home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow_internal.py", line 24, in swig_import_helper
    _mod = imp.load_module('_pywrap_tensorflow_internal', fp, pathname, description)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 243, in load_module
    return load_dynamic(name, filename, file)
  File "/home/pj/anaconda3/envs/tflow/lib/python3.6/imp.py", line 343, in load_dynamic
    return _load(spec)
ImportError: /home/pj/.cache/bazel/_bazel_pj/f508fdcc670cd71c05e44d5ea5f363d7/execroot/org_tensorflow/bazel-out/host/bin/tensorflow/python/keras/api/create_tensorflow.python_api_2_keras_python_api_gen_compat_v2.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN10tensorflow7functor16ScatterNdFunctorIN5Eigen16ThreadPoolDeviceESt7complexIfExLNS_13scatter_nd_op8UpdateOpE3ELi7EEclERKS3_xNS2_5arrayIlLm7EEENS2_9TensorMapINS2_6TensorIS5_Li2ELi1ElEELi16ENS2_11MakePointerEEENSD_INSE_IKxLi2ELi1ElEELi16ESG_EENSD_INSE_IKS5_Li2ELi1ElEELi16ESG_EESH_


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/errors

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.
Target //tensorflow/tools/api/tests:api_compatibility_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2402.569s, Critical Path: 2076.87s
INFO: 406 processes: 406 local.
FAILED: Build did NOT complete successfully
FAILED: Build did NOT complete successfully

```

pratjosh9 committed:
Remove unnecessary changes.

pratjosh9 committed:
Merge pull request #3 from pratjosh9/non_stateful_scatter_min_max

Remove unnecessary changes.

alextp commented:
@pratjosh9 your bazel tensorflow setup is borked. Did you run ./configure?

pratjosh9 committed:
Removed cases where complex datatype was being used for min/max

pratjosh9 committed:
Remove registration for ScatterNdMax/ScatterNdMin completely.

pratjosh9 committed:
Merge pull request #4 from pratjosh9/non_stateful_scatter_min_max

Non stateful scatter min max

pratjosh9 commented:
> @pratjosh9 your bazel tensorflow setup is borked. Did you run ./configure?

Yes. I configured it to compile with clang this time. 

In the file [scatter_nd_op_cpu_impl.h](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/scatter_nd_op_cpu_impl.h)

if I define the MIN op as 
```C++
template <typename Input, typename Update, typename Output>
class UpdateExecutor<Input, Update, Output, scatter_nd_op::UpdateOp::MIN> {
 public:
  EIGEN_STRONG_INLINE static void Execute(Input value/* input */, Update update,
                                          Output output) {
    output = value.cwiseMin(update);
  }
};
```
I get the error mentioned [here](https://github.com/tensorflow/tensorflow/pull/26923#issuecomment-485221991).



And, If I define the MIN op as 
```C++
template <typename Input, typename Update, typename Output>
class UpdateExecutor<Input, Update, Output, scatter_nd_op::UpdateOp::MIN> {
 public:
  EIGEN_STRONG_INLINE static void Execute(Input /* input */, Update update,
                                          Output output) {
    output = output.cwiseMin(update);
  }
};
```
I get the following error
```
./tensorflow/core/kernels/scatter_nd_op_cpu_impl.h:99:21: error: no member named 'cwiseMin' in 'Eigen::TensorDevice<Eigen::TensorChippingOp<0, Eigen::TensorMap<Eigen::Tensor<float, 2, 1, long>, 16, MakePointer> >, Eigen::ThreadPoolDevice>'
    output = output.cwiseMin(update);
             ~~~~~~ ^
./tensorflow/core/kernels/scatter_nd_op_cpu_impl.h:150:18: note: in instantiation of member function 'tensorflow::update_executor::UpdateExecutor<Eigen::TensorChippingOp<0, Eigen::TensorMap<Eigen::Tensor<float, 2, 1, long>, 16, MakePointer> >, Eigen::TensorChippingOp<0, const Eigen::TensorMap<Eigen::Tensor<const float, 2, 1, long>, 16, MakePointer> >, Eigen::TensorDevice<Eigen::TensorChippingOp<0, Eigen::TensorMap<Eigen::Tensor<float, 2, 1, long>, 16, MakePointer> >, Eigen::ThreadPoolDevice>, tensorflow::scatter_nd_op::UpdateOp::MIN>::Execute' requested here
            OP>::Execute(input_chip, update_chip, output_chip);
```

In the definition of [Eigen::TensorDevice](https://eigen.tuxfamily.org/dox/unsupported/TensorDevice_8h_source.html) it turns out that only +=, -= and the = operators are supported, using which the ops  `scatter_nd::UpdateOp ADD`, `scatter_nd::UpdateOp SUB` and `scatter_nd::UpdateOp ASSIGN` are implemented.

So how do I define the MIN and MAX functions with using cwiseMin() and cwiseMax() functions.

alextp commented:
@rmlarsen I think this goes beyond my knowledge of eigen; do you know what the right answer is here?

pratjosh9 commented:
Any updates?

rmlarsen commented:
Let me take a look...

pratjosh9 commented:
Is there anything with which I can help with in this issue?

rmlarsen commented:
@pratjosh9 This code has a little too much C++ magic for its own good. The functors should be updated to take both the output TensorMap and the device as separate arguments. So instead of 

      } else {
        auto input_chip = Toutput.template chip<0>(i);
        auto output_chip = input_chip.device(d);
        auto update_chip = Tupdates.template chip<0>(loc);
        update_executor::UpdateExecutor<
            decltype(input_chip), decltype(update_chip), decltype(output_chip),
            OP>::Execute(input_chip, update_chip, output_chip);
        ...


You would call   this as
        auto input_chip = Toutput.template chip<0>(i);
        auto output_chip = input_chip;
        auto update_chip = Tupdates.template chip<0>(loc);
        update_executor::UpdateExecutor<
            decltype(input_chip), decltype(update_chip), decltype(output_chip),
            OP>::Execute(input_chip, update_chip, output_chip);
      update_executor::UpdateExecutor<
            decltype(input_chip), decltype(update_chip), decltype(output_chip), Device,
            OP>::Execute(d, input_chip, update_chip, output_chip);


where the functor should be defined as:

template <typename DeviceType, typename Input, typename Update, typename Output>
class UpdateExecutor<DeviceType, Input, Update, Output, scatter_nd_op::UpdateOp::ADD> {
 public:
  EIGEN_STRONG_INLINE static void Execute(DeviceType device, Input /* input */, Update update,
                                          Output output) {
    output.device(device) += update;
  }
};

rmlarsen reviewed (changes requested):
See my comment in the conversation.

pratjosh9 commented:
Okay. Will start implementing these changes.

pratjosh9 committed:
Add DeviceType parameter to UpdateExecutor

pratjosh9 commented:
Any updates?

rmlarsen commented:
@pratjosh9 It looks like you are mixing device type enums and device types. For new ops you don't have to implement them for SYCL unless you want to, but most ops have such implementations and related instantiations.

mellanox-github commented:
Can one of the admins verify this patch?

pratjosh9 committed:
Fix bugs

pratjosh9 commented:
@rmlarsen I've made the requested changes to the functor. I'm not getting any of the previous errors but the build process hangs on my system so I couldn't generate the golden files. Are these changes fine?

Songminkee commented:
Is this an issue currently resolved?

rthadur commented:
@pratjosh9 can you please resolve conflicts ?

pratjosh9 committed:
Update scatter_nd_op_cpu_impl.h

pratjosh9 committed:
Update scatter_nd_op.cc

pratjosh9 committed:
Merge pull request #5 from pratjosh9/pratjosh9-patch-1

Fix Merge Conflicts

pratjosh9 committed:
Update scatter_nd_op.cc

pratjosh9 committed:
Update scatter_nd_op_cpu_impl.h

pratjosh9 commented:
> @pratjosh9 can you please resolve conflicts ?

Resolved

gbaned commented:
@pratjosh9  Could you please check failed build errors? Thanks!

pratjosh9 commented:
> @pratjosh9 Could you please check failed build errors? Thanks!

I'm removing declarations for min / max operations made for complex data types. 

Is there anyway to build and check these changes online? The build process on my local system is very slow and often lead to a non-responsive system. Thanks!

pratjosh9 committed:
Remove min/max declarations for complex data types

pratjosh9 committed:
Fix bug related to \ in macro expansion.

pratjosh9 committed:
Update scatter_nd_op_cpu_impl.h

pratjosh9 committed:
Merge pull request #6 from pratjosh9/pratjosh9-patch-1

Update scatter_nd_op_cpu_impl.h

pratjosh9 committed:
Merge remote-tracking branch 'upstream/master'

pratjosh9 committed:
Fix bug related to macro expansion.

pratjosh9 commented:
@rthadur @alextp Can anyone generate API goldens? The process is getting stuck on my laptop. The errors in the CPU build process were related to the goldens not being updated, and the errors in GPU build process were due to a typo in the name of a macro call and has been fixed.

rthadur commented:
@pratjosh9 sure will try to generate goldens internally, thank you

mihaimaruseac commented on tensorflow/python/kernel_tests/scatter_nd_ops_test.py:
```diff
@@ -96,6 +96,14 @@ def _NumpyDiv(ref, indices, updates):
 
 
+def _NumpyMin(ref, indices, updates):
+  return _NumpyScatterNd(ref, indices, updates, lambda p, u: min(p, u))
```
There is no need for lambda here, `lambda p, u: min(p, u)` is exactly the same as `min`.

pratjosh9 replied:
I've changed the call and removed the lambda function.

mihaimaruseac commented:
Also, can you please fix tests. It cannot be imported properly at the moment

gbaned commented:
@pratjosh9 Can you please check mihaimaruseac's comments and keep us posted? Thanks!

pratjosh9 committed:
Update scatter_nd_ops_test.py

Removed lambda function in the definition of _NumpyMin and _NumpyMax functions.

pratjosh9 commented:
> Also, can you please fix tests. It cannot be imported properly at the moment

The tests are failing because the API goldens are outdated. Unfortunately, I can't generate goldens as my current machine does not have the required compute power to build TensorFlow.

mihaimaruseac reviewed (approved):
Ok. I'll try to generate the API files inside

gbaned commented:
@mihaimaruseac Any update on this PR? Please. Thanks!

mihaimaruseac commented:
I imported the PR internally but there are ~5000 failing tests. Looking for a fix

mihaimaruseac commented:
So, the failure to generate the goldens is not because something being broken in the VM. Instead, this PR causes linker errors which manifest as `_ZN10tensorflow7functor16ScatterNdFunctorIN5Eigen16ThreadPoolDeviceESt7complexIfExLNS_13scatter_nd_op8UpdateOpE3ELi7EEclERKS3_xNS2_5arrayIlLm7EEENS2_9TensorMapINS2_6TensorIS5_Li2ELi1ElEELi16ENS2_11MakePointerEEENSD_INSE_IKxLi2ELi1ElEELi16ESG_EENSD_INSE_IKS5_Li2ELi1ElEELi16ESG_EESH_` (and similar not being found).

Probable reason (one of the many, you'll actually have to look closely at the build log and see more linker failures and fix them one by one):

```cc
bazel-out/k8-opt/bin/tensorflow/core/kernels/libscatter_nd_op.lo(scatter_nd_op.o): In function `tensorflow::Status tensorflow::functor::DoScatterNd<Eigen::GpuDevice, std::complex<double>, long long, (tensorflow::scatter_nd_op::UpdateOp)4>(tensorflow::OpKernelContext*, tensorflow::Tensor const&, tensorflow::Tensor const&, tensorflow::TensorShape const&, tensorflow::Tensor*, bool)':
scatter_nd_op.cc:(.text._ZN10tensorflow7functor11DoScatterNdIN5Eigen9GpuDeviceESt7complexIdExLNS_13scatter_nd_op8UpdateOpE4EEENS_6StatusEPNS_15OpKernelContextERKNS_6TensorESD_RKNS_11TensorShapeEPSB_b[_ZN10tensorflow7functor11DoScatterNdIN5Eigen9GpuDeviceESt7complexIdExLNS_13scatter_nd_op8UpdateOpE4EEENS_6StatusEPNS_15OpKernelContextERKNS_6TensorESD_RKNS_11TensorShapeEPSB_b]+0x319): undefined reference to `tensorflow::functor::ScatterNdFunctor<Eigen::GpuDevice, std::complex<double>, long long, (tensorflow::scatter_nd_op::UpdateOp)4, 7>::operator()(Eigen::GpuDevice const&, long long, Eigen::array<long, 7ul>, Eigen::TensorMap<Eigen::Tensor<std::complex<double>, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<long long const, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<std::complex<double> const, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<std::complex<double>, 2, 1, long>, 16, Eigen::MakePointer>)'
... (many many many more lines)
scatter_nd_op.cc:(.text._ZN10tensorflow7functor11DoScatterNdIN5Eigen9GpuDeviceESt7complexIfEiLNS_13scatter_nd_op8UpdateOpE3EEENS_6StatusEPNS_15OpKernelContextERKNS_6TensorESD_RKNS_11TensorShapeEPSB_b[_ZN10tensorflow7functor11DoScatterNdIN5Eigen9GpuDeviceESt7complexIfEiLNS_13scatter_nd_op8UpdateOpE3EEENS_6StatusEPNS_15OpKernelContextERKNS_6TensorESD_RKNS_11TensorShapeEPSB_b]+0x842): undefined reference to `tensorflow::functor::ScatterNdFunctor<Eigen::GpuDevice, std::complex<float>, int, (tensorflow::scatter_nd_op::UpdateOp)3, 6>::operator()(Eigen::GpuDevice const&, int, Eigen::array<long, 6ul>, Eigen::TensorMap<Eigen::Tensor<std::complex<float>, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<int const, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<std::complex<float> const, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<std::complex<float>, 2, 1, long>, 16, Eigen::MakePointer>)'
collect2: error: ld returned 1 exit status
```

Instead of building everything you can also try building smaller targets that depend on `scatter_nd_op`.

Internally we have nearly 5000 tests failing with linker errors such as

```cc
ld.lld: error: undefined symbol: tensorflow::functor::ScatterNdFunctor<Eigen::GpuDevice, std::__u::complex<float>, int, (tensorflow
::scatter_nd_op::UpdateOp)4, 6>::operator()(Eigen::GpuDevice const&, int, Eigen::array<long, 6ul>, Eigen::TensorMap<Eigen::Tensor<s
td::__u::complex<float>, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<Eigen::Tensor<int const, 2, 1, long>, 16, Eigen::Ma
kePointer>, Eigen::TensorMap<Eigen::Tensor<std::__u::complex<float> const, 2, 1, long>, 16, Eigen::MakePointer>, Eigen::TensorMap<E
igen::Tensor<std::__u::complex<float>, 2, 1, long>, 16, Eigen::MakePointer>)                                                       
>>> referenced by scatter_nd_op.cc                                                                                                 
>>>               blaze-out/k8-cuda101-opt/bin/third_party/tensorflow/core/kernels/_objs/scatter_nd_op/scatter_nd_op.o:(tensorflow:
:Status tensorflow::functor::DoScatterNd<Eigen::GpuDevice, std::__u::complex<float>, int, (tensorflow::scatter_nd_op::UpdateOp)4>(t
ensorflow::OpKernelContext*, tensorflow::Tensor const&, tensorflow::Tensor const&, tensorflow::TensorShape const&, tensorflow::Tens
or*, bool))  
```

mihaimaruseac commented:
Solved the linker error. Will be merged internally. Nothing to do here

mihaimaruseac commented on tensorflow/python/kernel_tests/scatter_nd_ops_test.py:
```diff
@@ -761,6 +780,56 @@ def testUpdateAddSubGradients(self):
       self.assertLess(err_subbed_wrt_updates, 2e-4)
 
+  @test_util.run_in_graph_and_eager_modes
+  def testUpdateMinMax(self):
```
This needs gradients to work. Removing from PR until gradients are added

mihaimaruseac commented on tensorflow/python/kernel_tests/scatter_nd_ops_test.py:
```diff
@@ -96,6 +96,14 @@ def _NumpyDiv(ref, indices, updates):
+
+
+def _NumpyMax(ref, indices, updates):
+  return _NumpyScatterNd(ref, indices, updates, max)
```
Inputs are numpy arrays so `max` is not proper here. Should be `np.maximum`. Handled manually

mihaimaruseac commented on tensorflow/python/kernel_tests/scatter_nd_ops_test.py:
```diff
@@ -232,6 +240,14 @@ def testVariableRankAdd(self):
+
+  @test_util.run_deprecated_v1
+  def testVariableRankMax(self):
+    self._VariableRankTests(_NumpyMax, state_ops.scatter_nd_max)
```
Test cannot be run on complex types, those should be removed from the test first

gbaned commented:
@pratjosh9 Can you please check @mihaimaruseac's comments and resolve conflicts?. Thanks!

mihaimaruseac commented:
This is handled internally. No need to change here (especially since changing will mean resolving merge conflicts first)

mihaimaruseac commented:
Thank you @pratjosh9 for the PR. It has been merged now.
